File: 3243-shortest-distance-after-road-addition-queries-i/3243-shortest-distance-after-road-addition-queries-i.py
from collections import deque, defaultdict

class Solution:
    def shortestDistanceAfterQueries(self, n, queries):
        adj = defaultdict(list)
        
        # Initialize the initial connections
        for i in range(1, n):
            adj[i - 1].append(i)
        
        def BFS(n):
            q = deque([0])
            dist = [float('inf')] * n
            dist[0] = 0
            
            while q:
                city = q.popleft()
                if city == n - 1:
                    return dist[city]
                
                for child in adj[city]:
                    if dist[child] > dist[city] + 1:
                        dist[child] = dist[city] + 1
                        q.append(child)
            
            return -1
        
        result = []
        for bridge in queries:
            adj[bridge[0]].append(bridge[1])
            result.append(BFS(n))
        
        return result

File: 3243-shortest-distance-after-road-addition-queries-i/3243-shortest-distance-after-road-addition-queries-i.cpp
class Solution {
public:
    vector<int>adj[500];
    int BFS(int n){
        queue<int>q; 
        q.push(0); // src,dist
        vector<int>dist(n,0x3f3f3f3f);
        dist[0] = 0;
        while(q.size()){
            int city = q.front();
            q.pop();
            if(city == n-1)return dist[city];
            for(int child : adj[city]){
                if(dist[child] > dist[city] +1){
                    dist[child] = dist[city] +1;
                    q.push(child);
                }
            }
        }
        return -1;
    }
    vector<int> shortestDistanceAfterQueries(int n, vector<vector<int>>& queries) {
        for(int i = 1 ; i < n;i++)
            adj[i-1].push_back(i);
        
        vector<int>ans;
        for(auto bridge : queries){ // Q
             adj[bridge[0]].push_back(bridge[1]); // O
             ans.push_back(BFS(n)); // E+V
        }
        return ans;
    }
};